src/primitives/sinsemilla.rs

```diff
@@ -64,59 +61,107 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
+/// [`sinsemilla::hash`] can be used.
+///
+/// [`sinsemilla::hash_to_point`]: self::hash_to_point
+/// [`sinsemilla::hash`]: self::hash
+pub trait HashDomain {
+    #[allow(non_snake_case)]
+    fn Q(&self) -> pallas::Point;
+}
+
+/// A domain with a specific prefix string, in which [`sinsemilla::hash_to_point`] and
+/// [`sinsemilla::hash`] can be used.
+///
+/// [`sinsemilla::hash_to_point`]: self::hash_to_point
+/// [`sinsemilla::hash`]: self::hash
+pub struct Domain(pub &'static str);
+
+impl HashDomain for Domain {
+    fn Q(&self) -> pallas::Point {
+        pallas::Point::hash_to_curve(Q_PERSONALIZATION)(self.0.as_bytes())
+    }
 }
 
 /// `SinsemillaHashToPoint` from [§ 5.4.1.9][concretesinsemillahash].
 ///
 /// [concretesinsemillahash]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
 #[allow(non_snake_case)]
-pub(crate) fn hash_to_point(domain_prefix: &str, msg: impl Iterator<Item = bool>) -> pallas::Point {
+pub(crate) fn hash_to_point<D: HashDomain>(
+    domain: &D,
+    msg: impl Iterator<Item = bool>,
+) -> pallas::Point {
     let padded: Vec<_> = Pad::new(msg).collect();
 
-    let hasher_S = pallas::Point::hash_to_curve(GROUP_HASH_S);
+    let hasher_S = pallas::Point::hash_to_curve(S_PERSONALIZATION);
     let S = |chunk: &[bool]| hasher_S(&lebs2ip_k(chunk).to_le_bytes());
 
     padded
         .chunks(K)
-        .fold(Q(domain_prefix), |acc, chunk| acc.double() + S(chunk))
+        .fold(domain.Q(), |acc, chunk| acc.double() + S(chunk))
 }
 
 /// `SinsemillaHash` from [§ 5.4.1.9][concretesinsemillahash].
 ///
 /// [concretesinsemillahash]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
-pub(crate) fn hash(domain_prefix: &str, msg: impl Iterator<Item = bool>) -> pallas::Base {
-    extract_p(&hash_to_point(domain_prefix, msg))
+pub(crate) fn hash<D: HashDomain>(domain: &D, msg: impl Iterator<Item = bool>) -> pallas::Base {
+    extract_p(&hash_to_point(domain, msg))
+}
+
+/// Trait representing a domain in which [`sinsemilla::commit`] and
+/// [`sinsemilla::short_commit`] can be used.
+///
+/// [`sinsemilla::commit`]: self::commit
+/// [`sinsemilla::short_commit`]: self::short_commit
+pub trait CommitDomain: HashDomain {
+    #[allow(non_snake_case)]
+    fn R(&self) -> pallas::Point;
+}
+
+/// A domain with a specific prefix string, in which [`sinsemilla::commit`] and
+/// [`sinsemilla::short_commit`] can be used.
+///
+/// [`sinsemilla::commit`]: self::commit
+/// [`sinsemilla::short_commit`]: self::short_commit
+pub struct Comm(pub &'static str);
+
+impl HashDomain for Comm {
+    fn Q(&self) -> pallas::Point {
+        let m_prefix = self.0.to_owned() + "-M";
+        pallas::Point::hash_to_curve(Q_PERSONALIZATION)(m_prefix.as_bytes())
+    }
+}
+
+impl CommitDomain for Comm {
+    fn R(&self) -> pallas::Point {
+        let r_prefix = self.0.to_owned() + "-r";
+        let hasher_r = pallas::Point::hash_to_curve(&r_prefix);
+        hasher_r(&[])
+    }
 }
```

Comment:
```suggestion
impl CommitDomain {
    fn new(domain: &str) -> Self {
        let m_prefix = domain.to_owned() + "-M";
        let r_prefix = domain.to_owned() + "-r";
        CommitDomain {
            Q: pallas::Point::hash_to_curve(Q_PERSONALIZATION)(m_prefix.as_bytes()),
            R: pallas::Point::hash_to_curve(r_prefix.as_bytes())(&[]),
        }
    }
}
```

Reply:
I'd tweak the above to `Q: HashDomain::new(m_prefix.as_bytes()).Q` which better reflects the relationship.